### PR TITLE
Multiz Rework: Human Suffering Edition (Contains PLANE CUBE)

### LINKED ISSUE #65800: Emissives show through walls on the Z-Level above them
Reporting client version: 514.1581

## Round ID:
[180764](https://scrubby.melonmesa.com/round/180764)

## Testmerges:
- [Pingbus changes and improvements](https://www.github.com/tgstation/tgstation/pull/65625)

## Reproduction:

Have an emissive (Air Alarm / Fire Alarm)
Place it on a wall
Go to the Z-Level above the wall, and place another wall.

Observe the following:

Upper Z-Level:

image

Lower Z-Level:

image

Those grey spots perfectly correspond to the emissive portion of those alarms, so I blame that. I think this only shows up when you have /turf/openspace present (since those two are in that little stairwell in the IceBox Kitchen Area).

### LINKED ISSUE #68461: Glass floors are made of pure suffering
## Reproduction:
Exist, really, thats about it

image

### PR DESCRIPTION
**Pull request by LemonInTheDark**
## About The Pull Request

I've reworked multiz. This was done because our current implementation of multiz flattens planes down into just the openspace plane. This breaks any effects we attach to plane masters (including lighting), but it also totally kills the SIDE_MAP map format, which we NEED for wallening (A major 3/4ths resprite of all wall and wall adjacent things, making them more then one tile high. Without sidemap we would be unable to display things both in from of and behind objects on map. Stupid.)

This required MASSIVE changes. Both to all uses of the plane var for reasons I'll discuss later, and to a ton of different systems that interact with rendering.

I'll do my best to keep this compact, but there's only so much I can do. Sorry brother.

### Core idea

OK: first thing.
`vis_contents` as it works now squishes the planes of everything inside it down into the plane of the vis_loc.
This is bad. But how to do better?

It's trivially easy to make copies of our existing plane masters but offset, and relay them to the bottom of the plane above. Not a problem. The issue is how to get the actual atoms on the map to "land" on them properly.

We could use `FLOAT_PLANE` to offset planes based off how they're being seen, in theory this would allow us to create lens for how objects are viewed.
But that's not a stable thing to do, because properly "landing" a plane on a desired plane master would require taking into account every bit of how it's being seen, would inherently break this effect.

Ok so we need to manually edit planes based off "z layer" (IE: what layer of a z stack are you on).

That's the key conceit of this pr. Implementing the plane cube, and ensuring planes are always offset properly.
Everything else is just gravy.

### About the Plane Cube 

Each plane master (except ones that opt out) is copied down by some constant value equal to the max absolute change between the first and the last plane.
We do this based off the max z stack size detected by SSmapping. This is also where updates come from, and where all our updating logic will live.

As mentioned, plane masters can choose to opt out of being mirrored down. In this case, anything that interacts with them assuming that they'll be offset will instead just get back the valid plane value. This works for render targets too, since I had to work them into the system as well.

Plane masters can also be temporarily hidden from the client's screen. This is done as an attempt at optimization, and applies to anything used in niche cases, or planes only used if there's a z layer below you.

### About Plane Master Groups

BYOND supports having different "maps" on screen at once (IE: groups of items/turfs/etc)
Plane masters cannot cover 2 maps at once, since their location is determined by their screen_loc.
So we need to maintain a mirror of each plane for every map we have open.

This was quite messy, so I've refactored it (and maps too) to be a bit more modular.

Rather then storing a list of plane masters, we store a list of plane master group datums.
Each datum is in charge of the plane masters for its particular map, both creating them, and managing them.

Like I mentioned, I also refactored map views. Adding a new mapview is now as simple as newing a `/atom/movable/screen/map_view`, calling `generate_view` with the appropriate map id, setting things you want to display in its `vis_contents`, and then calling `display_to` on it, passing in the mob to show ourselves to.

Much better then the hardcoded pattern we used to use. So much duplicated code man.

Oh and plane master controllers, that system we have that allows for applying filters to sets of plane masters? I've made it use lookups on plane master groups now, rather then hanging references to all impacted planes. This makes logic easier, and prevents the need to manage references and update the controllers.

image

In addition, I've added a debug ui for plane masters.
It allows you to view all of your own plane masters and short descriptions of what they do, alongside tools for editing them and their relays.

It ALSO supports editing someone elses plane masters, AND it supports (in a very fragile and incomplete manner) viewing literally through someone else's eyes, including their plane masters. This is very useful, because it means you can debug "hey my X is yorked" issues yourself, on live.

In order to accomplish this I have needed to add setters for an ungodly amount of visual impacting vars. Sight flags, eye, see_invis, see_in_dark, etc.

It also comes with an info dump about the ui, and plane masters/relays in general.

Sort of on that note. I've documented everything I know that's niche/useful about our visual effects and rendering system. My hope is this will serve to bring people up to speed on what can be done more quickly, alongside making my sin here less horrible.
See https://github.com/LemonInTheDark/tgstation/blob/multiz-hell/.github/guides/VISUALS.md.

### "Landing" planes

Ok so I've explained the backend, but how do we actually land planes properly?
Most of the time this is really simple. When a plane var is set, we need to provide some spokesperson for the appearance's z level. We can use this to derive their z layer, and thus what offset to use.

This is just a lot of gruntwork, but it's occasionally more complex.
Sometimes we need to cache a list of z layer -> effect, and then use that.
Also a LOT of updating on z move. So much z move shit.

Oh. and in order to make byond darkness work properly, I needed to add SEE_BLACKNESS to all sight flags.
This draws darkness to plane 0, which means I'm able to relay it around and draw it on different z layers as is possible. ~~fun darkness ripple effects incoming someday~~

I also need to update mob overlays on move.
I do this by realiizing their appearances, mutating their plane, and then readding the overlay in the correct order.

The cost of this is currently 3N. I'm convinced this could be improved, but I've not got to it yet.
It can also occasionally cause overlays to corrupt. This is fixed by laying a protective ward of `overlays.Copy` in the sand, but that spell makes the compiler confused, so I'll have to bully lummy about fixing it at some point.

### Behavior changes

We've had to give up on the already broken gateway "see through" effect. Won't work without managing gateway plane masters or something stupid. Not worth it.
So instead we display the other side as a ui element. It's worse, but not that bad.

Because vis_contents no longer flattens planes (most of the time), some uses of it now have interesting behavior.
The main thing that comes to mind is alert popups that display mobs. They can impact the lighting plane.
I don't really care, but it should be fixable, I think, given elbow grease.

Ah and I've cleaned up layers and plane defines to make them a bit easier to read/reason about, at least I think.

## Why It's Good For The Game

Behold, you can see multi multizs 
image
image
Also dropshadow and lighting works now
image
image

Fixes #65800
Fixes #68461 

## Changelog

:cl:
refactor: Refactored... well a lot really. Map views, anything to do with planes, multiz, a shit ton of rendering stuff. Basically if you see anything off visually report it
admin: VV a mob, and hit View/Edit Planes in the dropdown to steal their view, and modify it as you like. You can do the same to yourself using the Edit/Debug Planes verb
/:cl:

### CODE REVIEW

**Comment by LemonInTheDark** [comment]
Paging @Kylerace and @timothymtorres since they both seemed to understand what I was trying to do here

**Comment by optimumtact** [comment]
Probably the most important change since Intialize dropped, great work

**Comment by optimumtact** [comment]
>The issue is how to get the actual atoms on the map to "land" on them properly.

NB: you don't actually explain how this is done, you hook creation and movement to place them into the appropriate planes based on z?

**Comment by optimumtact** [comment]
also, how do openspace tiles know which plane to render as their contents?

**Comment by MrStonedOne** [comment]
>AND it supports (in a very fragile and incomplete manner) viewing literally through someone else's eyes, including their plane masters. This is very useful, because it means you can debug "hey my X is yorked" issues yourself, on live.

This already exists on master.

The `observe` verb as a ghost will let you do this to any mob. View thru their hud, which controls their planemasters.

Did you preserve this?

Was your getters/setters work a duplication of the hud datum initialization system?

**Comment by LemonInTheDark** [comment]
> NB: you don't actually explain how this is done, you hook creation and movement to place them into the appropriate planes based on z?

Whenever a plane variable is set (or on atom initialize/z move), I require some form of "spokesman" that can be used to get_turf the z level of the thing being set on. (outside niche cases)
We store a list of z level (as number) -> z level offset
Whenever a new z level is loaded, we check to see what level of offset it's in. the top level is 0, down 1 is 1, etc.

Then when we attempt to set plane, we index into that list on SSmapping with the z level, and get handed back the offset to use
Then it's just a matter of subtracting (offset * max plane delta), and we're done, things will land properly.

> also, how do openspace tiles know which plane to render as their contents?

Openspace tiles use vis_contents as before to tell byond to attempt to render the turfs below
The turfs below are already on lower turfs, so when they are rendered they'll be piped up through the set of render relays, onto their level's master plate, and then onto the transparent plane, so they'll be rendered visually "below"

> This already exists on master.
> 
> The `observe` verb as a ghost will let you do this to any mob. View thru their hud, which controls their planemasters.
> 
> Did you preserve this?
> 
> Was your getters/setters work a duplication of the hud datum initialization system?

the observe verb doesn't use the plane masters of the observed mob. I do.
This makes my code much more fragile, but lets me actually debug PM issues.
It ought? to all work, though I think what I did to sight might have yorked it slightly, will fix.

I don't think the hud datum actually tracks this stuff, so no?

**Comment by LemonInTheDark** [comment]
I've added support for disabling temporary/limited (think ai static, pipecrawl, etc) plane masters. This hides them and their relays from our client screen, hopefully reducing client load.

This hides between 13% and 20% of our PM count, depending on the offset count.

I'm gonna look into adding a low specs pref and using it to disable things like filters on out of view planes. Not sure how large an impact this would actually make however.

**Comment by LemonInTheDark** [comment]
Might not actually we don't have all that many filters and I'm not sure how big an impact it would actually make.
Would need to see people having issues before I dug that deep.

**Comment by Time-Green** [comment]
Multi-z levels look a bit too dark
image

image

**Comment by LemonInTheDark** [comment]
Yeet
it's a consequence of how we signify depth, that little dark bit we overlay on multiz
I'm working on a different multiz implementation so we can drop that, since it only worked before because lighting was yorked

File: code/modules/paperwork/paperbin.dm [hunk 1]
```diff
@@ -130,23 +130,21 @@
 
 /obj/item/paper_bin/examine(mob/user)
 	. = ..()
+	var/total_paper = LAZYLEN(papers)
```
Comment (ShizCalev) [inline code/modules/paperwork/paperbin.dm:133]: this doesn't seem relevant to this PR?

File: code/modules/paperwork/paperbin.dm [hunk 2]
```diff
@@ -14,7 +14,7 @@
 	throw_range = 7
 	pressure_resistance = 8
 	var/papertype = /obj/item/paper
-	var/total_paper = 30
+	var/initial_paper = 30
```
Comment (ShizCalev) [inline code/modules/paperwork/paperbin.dm:17]: .

File: code/modules/paperwork/paperbin.dm [hunk 3]
```diff
@@ -33,8 +33,8 @@
 		if(pen && !bin_pen)
 			pen.forceMove(src)
 			bin_pen = pen
-	for(var/i in 1 to total_paper)
-		papers.Add(generate_paper())
+	for(var/i in 1 to initial_paper)
+		papers += generate_paper()
```
Comment (ShizCalev) [inline code/modules/paperwork/paperbin.dm:37]: .

File: code/modules/paperwork/paperbin.dm [hunk 4]
```diff
@@ -130,23 +130,21 @@
...
 	. = ..()
+	var/total_paper = LAZYLEN(papers)
 	if(total_paper)
 		. += "It contains [total_paper > 1 ? "[total_paper] papers" : "one paper"]."
 	else
 		. += "It doesn't contain anything."
 
 /obj/item/paper_bin/update_icon_state()
-	if(total_paper < 1)
+	if(!LAZYLEN(papers))
```
Comment (ShizCalev) [inline code/modules/paperwork/paperbin.dm:140]: .

File: code/modules/paperwork/paperplane.dm [hunk 5]
```diff
@@ -57,7 +57,7 @@
 
 /obj/item/paperplane/update_overlays()
 	. = ..()
-	for(var/stamp in internalPaper.stamp_cache)
+	for(var/stamp in internalPaper?.stamp_cache)
```
Comment (ShizCalev) [inline code/modules/paperwork/paperplane.dm:60]: .

File: _maps/RandomRuins/SpaceRuins/onehalf.dmm [hunk 6]
```diff
@@ -680,7 +680,7 @@
 	},
 /obj/structure/table/reinforced,
 /obj/item/paper_bin{
-	total_paper = 12
+	initial_paper = 12
```
Comment (ShizCalev) [inline _maps/RandomRuins/SpaceRuins/onehalf.dmm:683]: .

File: _maps/shuttles/whiteship_donut.dmm [hunk 7]
```diff
@@ -399,7 +399,7 @@
 /obj/structure/table,
 /obj/effect/decal/cleanable/dirt/dust,
 /obj/item/paper_bin{
-	total_paper = 20
+	initial_paper = 20
```
Comment (ShizCalev) [inline _maps/shuttles/whiteship_donut.dmm:402]: .

File: code/game/objects/items/devices/portable_chem_mixer.dm [hunk 8]
```diff
@@ -67,7 +67,7 @@
 	return
 
 /obj/item/storage/portable_chem_mixer/update_icon_state()
-	if(!atom_storage.locked)
+	if(!atom_storage?.locked)
```
Comment (ShizCalev) [inline code/game/objects/items/devices/portable_chem_mixer.dm:70]: ?

File: code/game/objects/items/cards_ids.dm [hunk 9]
```diff
@@ -111,8 +113,8 @@
 	registered_account.replaceable = TRUE
 
 	// Applying the trim updates the label and icon, so don't do this twice.
-	if(ispath(trim))
-		SSid_access.apply_trim_to_card(src, trim)
+	if(ispath(trim_path) && !trim)
+		SSid_access.apply_trim_to_card(src, trim_path)
```
Comment (ShizCalev) [inline code/game/objects/items/cards_ids.dm:117]: what does this have to do with multiz

File: code/modules/paperwork/paperbin.dm (hunk 1 above)
Comment (LemonInTheDark) [reply code/modules/paperwork/paperbin.dm:133]: one would think
I setup code that calls update_overlays on init, for reasons, and since paperbins used the var this replaced as both a cache of the paper count AND a way to tell themselves on init how much paper to spawn, it broke
Fuck I forgot to mention that in the pr. I will.

https://github.com/tgstation/tgstation/pull/67902#issuecomment-1191952086

File: code/game/objects/items/devices/portable_chem_mixer.dm (hunk 8 above)
Comment (LemonInTheDark) [reply code/game/objects/items/devices/portable_chem_mixer.dm:70]: relevant to above, this can go null

File: code/game/objects/items/cards_ids.dm (hunk 9 above)
Comment (LemonInTheDark) [reply code/game/objects/items/cards_ids.dm:117]: https://github.com/LemonInTheDark/tgstation/commit/644813af784dc9b5a900e2d4d3f6dfad176fc54f

Comment (LemonInTheDark) [reply code/game/objects/items/cards_ids.dm:117]: Caused runtimes in init, and is also just annoying

File: code/modules/paperwork/paperbin.dm (hunk 1 above)
Comment (LemonInTheDark) [reply code/modules/paperwork/paperbin.dm:133]: https://github.com/LemonInTheDark/tgstation/commit/644813af784dc9b5a900e2d4d3f6dfad176fc54f

File: code/game/objects/structures/headpike.dm [hunk 10]
```diff
@@ -40,6 +40,9 @@
 	return ..()
 
 /obj/structure/headpike/update_name()
+	if(!victim)
+		name = initial(name)
+		return ..()
 	name = "[victim.real_name] on a [spear]"
```
Comment (ShizCalev) [inline code/game/objects/structures/headpike.dm:46]: name = victim ? "[victim.real_name] on a [spear]" : initial(name)

File: code/game/machinery/lightswitch.dm [hunk 11]
```diff
@@ -40,19 +40,19 @@ MAPPING_DIRECTIONAL_HELPERS(/obj/machinery/light_switch, 26)
 	luminosity = (machine_stat & NOPOWER) ? 0 : 1
 
 /obj/machinery/light_switch/update_icon_state()
-	set_light(area.lightswitch ? 0 : light_on_range)
+	set_light(area?.lightswitch ? 0 : light_on_range)
```
Comment (ShizCalev) [inline code/game/machinery/lightswitch.dm:43]: @optimumtact iirc you yelled at me two years ago for trying to add this exact check lol

I assume this is encountered during initialize / destroy?

File: code/game/objects/items/tools/weldingtool.dm [hunk 12]
```diff
@@ -190,7 +190,7 @@
 
 /// Returns the amount of fuel in the welder
 /obj/item/weldingtool/proc/get_fuel()
-	return reagents.get_reagent_amount(/datum/reagent/fuel)
+	return reagents?.get_reagent_amount(/datum/reagent/fuel)
```
Comment (ShizCalev) [inline code/game/objects/items/tools/weldingtool.dm:193]: Honestly I'd prefer explicitly checking if !initialized || qdeleting and returning on that since this is more likely to end up hiding different problems

File: code/game/objects/items/cards_ids.dm (hunk 9 above)
Comment (ShizCalev) [inline code/game/objects/items/cards_ids.dm:117]: I'm guessing this runtime that I've been seeing on sybil is what this fixes? Been eyeing it for a few days

```
[2022-08-12 09:53:52.419] runtime error: Cannot read /datum/id_trim/job/prisoner (/datum/id_trim/job/prisoner).sechud_icon_state
 - proc name: get trim sechud icon state (/obj/item/card/id/proc/get_trim_sechud_icon_state)
 -   source file: cards_ids.dm,790
 -   usr: Helvetica8563 (/mob/dead/new_player)
 -   src: the prisoner ID card (/obj/item/card/id/advanced/prisoner)
 -   usr.loc: the plating (8,174,1) (/turf/open/floor/plating)
 -   src.loc: Haliee Mason (/mob/living/carbon/human/dummy)
 -   call stack:
 - the prisoner ID card (/obj/item/card/id/advanced/prisoner): get trim sechud icon state()
 - the prisoner ID card (/obj/item/card/id/advanced/prisoner): get trim sechud icon state()
 - the prisoner ID card (/obj/item/card/id/advanced/prisoner): get sechud job icon state()
 - Haliee Mason (/mob/living/carbon/human/dummy): sec hud set ID()
 - Haliee Mason (/mob/living/carbon/human/dummy): equip to slot(the prisoner ID card (/obj/item/card/id/advanced/prisoner), 256, 1, 0)
 - Haliee Mason (/mob/living/carbon/human/dummy): equip to slot if possible(the prisoner ID card (/obj/item/card/id/advanced/prisoner), 256, 1, 1, 0, 1, 1)
 - Haliee Mason (/mob/living/carbon/human/dummy): equip to slot or del(the prisoner ID card (/obj/item/card/id/advanced/prisoner), 256, 1)
 - Prisoner (/datum/outfit/job/prisoner): equip(Haliee Mason (/mob/living/carbon/human/dummy), 1)
 - Haliee Mason (/mob/living/carbon/human/dummy): equipOutfit(/datum/outfit/job/prisoner (/datum/outfit/job/prisoner), 1)
 - Haliee Mason (/mob/living/carbon/human/dummy): dress up as job(/datum/job/prisoner (/datum/job/prisoner), 1)
 - ...
 - /datum/preferences (/datum/preferences): create character preview view(Helvetica8563 (/mob/dead/new_player))
 - /datum/preferences (/datum/preferences): ui data(Helvetica8563 (/mob/dead/new_player))
 - /datum/tgui (/datum/tgui): get payload(null, 1, 1)
 - /datum/tgui (/datum/tgui): open()
 - /datum/preferences (/datum/preferences): ui interact(Helvetica8563 (/mob/dead/new_player), /datum/tgui (/datum/tgui))
 -  (/atom/movable/screen/lobby/button/character_setup): Click(null, "mapwindow.map", "icon-x=61;icon-y=21;left=1;but...")
 -  (/atom/movable/screen/lobby/button/character_setup):  Click(null, "mapwindow.map", "icon-x=61;icon-y=21;left=1;but...")
 - /datum/callback/verb_callback (/datum/callback/verb_callback): Invoke()
 - /datum/callback/verb_callback (/datum/callback/verb_callback): InvokeAsync()
 - 
```

Comment (LemonInTheDark) [reply code/game/objects/items/cards_ids.dm:117]: Likely yes. trim stored both the path, and the object created from the path. the bug there could just be trying to get icon state before the trim is new'd, but I'm not sure when that ordinarily happens

File: code/game/machinery/lightswitch.dm (hunk 11 above)
Comment (optimumtact) [reply code/game/machinery/lightswitch.dm:43]: u little snitch lol

it seems very unlikely area would ever be null in normal situations, so I still don't agree with the nullability check

Comment (LemonInTheDark) [reply code/game/machinery/lightswitch.dm:43]: I thought about it a bit, I think it'd be better to check for the initialized flag in gases like this 
since we don't want these cases running until we're out of initialize, we should just make that explicit

File: code/game/objects/structures/headpike.dm (hunk 10 above)
Comment (LemonInTheDark) [reply code/game/objects/structures/headpike.dm:46]: idk, if you're sure. I'm not a huge fan of ternaries but maybe it's fine here

File: code/game/machinery/lightswitch.dm (hunk 11 above)
Comment (LemonInTheDark) [reply code/game/machinery/lightswitch.dm:43]: god fucking damn it I can't. SHIT

Comment (LemonInTheDark) [reply code/game/machinery/lightswitch.dm:43]: This sort of thing is really annoying because it's "this breaks if called before init is **mostly** done"
we need to be able to call it IN init, but we can't run it before area has been set, because if we do we get runtimes.
I can't even use the flag because the call I'm handing happens AFTER the flag is set

help

**Comment by timothymtorres** [comment]
Possible issues this might solve:

- #64385
- #64098
- #62442
- #61968
- #59610
- #56390
- #53783
- #53531
- #50444
- #49890
- #45669
- #40542
- #39799
- #67336
- #37303
- #53581
- #68461
- #68154

**Comment by msgerbs** [comment]
Could this also potentially fix this pesky bastard?

* [ Welding masks intermittently not blocking vision #52972](https://github.com/tgstation/tgstation/issues/52972)

**Comment by LemonInTheDark** [comment]
I'm sorry but I'm gonna really disappoint you, I don't think I've hit most of these

* [#64279 layered borg eyelights above mobs and cover-overlays #64385](https://github.com/tgstation/tgstation/issues/64385) Nope, likely a consequence of eyelights using the above lighting plane rather them emissives

* [Emissives do not change when the wearer lays down #64098](https://github.com/tgstation/tgstation/issues/64098) See below

* [emissive blockers aren't applied properly for both body parts and worn overlays #62442](https://github.com/tgstation/tgstation/issues/62442) Also nope, this is likely just the reset_transform fuckin us up

* [Overlays from BCI going under tiles(?) #61968](https://github.com/tgstation/tgstation/issues/61968) honestly no fucking idea what's wrong here, outside of those things not being on the hud plane. what even ARE they can't find em in the code

* [Cryo tubes don't always seem to get rid of the occupant (overlay?) when the occupant is ejected #59610](https://github.com/tgstation/tgstation/issues/59610) I could swear we fixed this already, I guess not

* [Skyburger flattened into singularity #56390](https://github.com/tgstation/tgstation/issues/56390) this isn't really a bug, tho someone do please make really tall burgers shake and have an approaching 100% chance to fall over as they get taller

* [Mesons allow seeing objects, not just turfs, in the z-level below you through walls #53783](https://github.com/tgstation/tgstation/issues/53783) This would be potentially fixable now that we have control over the blackness plane, but I've not done so yet

* [Radiation storm overlay interferes with runechat #53531](https://github.com/tgstation/tgstation/issues/53531) I may have fixed this with my weather effects pr, honestly not sure how it was happening in the first place. doesn't happen now tho.

* [When Ventcrawling Everything Is Dark. #50444](https://github.com/tgstation/tgstation/issues/50444) I believe this is because slimes have a low see_invisible value, which masks away the fullbright and fulldark images. not totally sure, but that's my assumption. not fixed here

* [Mob overlays blur when mobs are resized #49890](https://github.com/tgstation/tgstation/issues/49890) Local man doesn't place image in issue report. Not sure what this means, as scaling a human works fine, and they are all overlay. needs more detail, not fixed by me

* [Some screens render over Nar'Sie #45669](https://github.com/tgstation/tgstation/issues/45669) I think the fix for this would be making narsie block emissives? Not sure

* [Mining scanners with mesons during ash storms leave "empty" tiles in place of unexposed ores instead of "rock" tiles #40542](https://github.com/tgstation/tgstation/issues/40542) ... WHAT? I cannot replicate this.

* [You can see items through walls with engineering meson scanner. #39799](https://github.com/tgstation/tgstation/issues/39799) We use an image for this, with a location of the target turf. This might be fixed just by using an overlay instead. Not sure

* [The light / emissive effect of beams are not aligned properly #67336](https://github.com/tgstation/tgstation/issues/67336) Nope. Same bug as before, emisssives refuse to rotate. Why? I have no fucking clue

* [Images on guns don't rotate when thrown #37303](https://github.com/tgstation/tgstation/issues/37303) This is interesting, they do rotate, but not in the same way as the gun, they refuse to turn. Not fixed here

* [Wall Smoothing issues on MultiZ #53581](https://github.com/tgstation/tgstation/issues/53581) Fixed yes. I want to redo how this works actually, the transparent element should be used here but it's been hacked to work for glass walls, so I'll need to fix that first. Maybe later, I wanna be able to see under top level shuttles damn it. 

* [Glass floors are made of pure suffering #68461](https://github.com/tgstation/tgstation/issues/68461) Fixed. Honestly not sure why? maybe it was the vis_contents rules of the gas overlay. But fixed.

* [alien resin sometimes doesn't change colors during timestop? #68154](https://github.com/tgstation/tgstation/issues/68154) Nope. I have no idea what would cause this, makes me think something weird with atom color's goin on

File: code/game/machinery/lightswitch.dm (hunk 11 above)
Comment (LemonInTheDark) [reply code/game/machinery/lightswitch.dm:43]: Resolved, I didn't actually need to call update_appearance

File: code/modules/industrial_lift/industrial_lift.dm [hunk 13]
```diff
@@ -17,6 +17,8 @@ GLOBAL_LIST_EMPTY(lifts)
 	canSmoothWith = list(SMOOTH_GROUP_INDUSTRIAL_LIFT)
 	obj_flags = CAN_BE_HIT | BLOCK_Z_OUT_DOWN
 	appearance_flags = PIXEL_SCALE|KEEP_TOGETHER //no TILE_BOUND since we're potentially multitile
+	// If we don't do this, we'll build our overlays early, and fuck up how we're rendered
+	blocks_emissive = 0
```
Comment (optimumtact) [inline code/modules/industrial_lift/industrial_lift.dm:21]: NONE

Comment (LemonInTheDark) [reply code/modules/industrial_lift/industrial_lift.dm:21]: aight

**Comment by LemonInTheDark** [comment]
I still have some stuff I need to work out with this, outside the obvious. space looks bad right now, and I'm worried about init times
Also uh, full transparency, I added parallaxing to multiz, and multiz lighting. things looked like shit without it, and it made me sad

if that's an issue to bundle with this I can atomize it out or whatever

**Comment by LemonInTheDark** [comment]
I think init is fine with this. We gained maybe 0.5 seconds on space init? but nothing that bad, that much was slightly expected

**Comment by LemonInTheDark** [comment]
TM'd on sybil, just to be sure

**Comment by ShizCalev** [comment]
```
[14:40:05] Runtime in stack_trace.dm, line 3: No plane offset passed in as context for a non floating mutable appearance, ya done fucked up
proc name: stack trace (/proc/stack_trace)
usr: BNQORD/(Joe Bicknell)
usr.loc: (Fore Primary Hallway (100,154,4))
src: null
call stack:
stack trace("No plane offset passed in as c...")
mutable appearance('icons/obj/clothing/suits/ether...', "eth_raincoat_glow", -1, null, 14, 255, 106, null)
emissive appearance('icons/obj/clothing/suits/ether...', "eth_raincoat_glow", null, -1, 255, 0, null)
the ethereal raincoat (/obj/item/clothing/suit/hooded/ethereal_raincoat): update overlays(16777215)
the ethereal raincoat (/obj/item/clothing/suit/hooded/ethereal_raincoat): update icon(16777215)
the ethereal raincoat (/obj/item/clothing/suit/hooded/ethereal_raincoat): update appearance(16777215)
the ethereal raincoat (/obj/item/clothing/suit/hooded/ethereal_raincoat): on changed z level(the plating (99,154,3) (/turf/open/floor/plating), the floor (100,154,4) (/turf/open/floor/iron), 0, 1)
Joe Bicknell (/mob/living/carbon/human): on changed z level(the plating (99,154,3) (/turf/open/floor/plating), the floor (100,154,4) (/turf/open/floor/iron), 0, 1)
Joe Bicknell (/mob/living/carbon/human): on changed z level(the plating (99,154,3) (/turf/open/floor/plating), the floor (100,154,4) (/turf/open/floor/iron), 0, null)
Joe Bicknell (/mob/living/carbon/human): on changed z level(the plating (99,154,3) (/turf/open/floor/plating), the floor (100,154,4) (/turf/open/floor/iron), 0, null)
...
the plating (99,154,3) (/turf/open/floor/plating): SendSignal("atom_exit", /list (/list))
Joe Bicknell (/mob/living/carbon/human): Move(the wall (100,154,3) (/turf/closed/wall), 4, 0)
Joe Bicknell (/mob/living/carbon/human): Move(the wall (100,154,3) (/turf/closed/wall), 4, null)
Joe Bicknell (/mob/living/carbon/human): Move(the wall (100,154,3) (/turf/closed/wall), 4)
Joe Bicknell (/mob/living/carbon/human): Move(the wall (100,154,3) (/turf/closed/wall), 6, null)
Joe Bicknell (/mob/living/carbon/human): Move(the wall (100,154,3) (/turf/closed/wall), 6)
Joe Bicknell (/mob/living/carbon/human): keyLoop(BNQORD (/client))
Input (/datum/controller/subsystem/verb_manager/input): ignite(0)
Master (/datum/controller/master): Loop(2)
Master (/datum/controller/master): StartProcessing(0)
```

**Comment by scriptis** [comment]
Camera consoles appear to be extremely broken, uncertain if it's related to this.

**Comment by LemonInTheDark** [comment]
would totally be related, thank you. will look into it

**Comment by ShizCalev** [comment]
```
[18:28:04] Runtime in stack_trace.dm, line 3: No plane offset passed in as context for a non floating mutable appearance, ya done fucked up
proc name: stack trace (/proc/stack_trace)
usr: Normmacdonaldfan/(Silent Retard)
usr.loc: (Starboard Primary Hallway (142,110,2))
src: null
call stack:
stack trace("No plane offset passed in as c...")
mutable appearance('icons/effects/effects.dmi', "thought_bubble", -1, null, 5, 255, 64, null)
Silent Retard (/mob/living/carbon/human): create point bubble(the pen (/obj/item/pen))
Silent Retard (/mob/living/carbon/human): point at(the pen (/obj/item/pen))
Silent Retard (/mob/living/carbon/human): pointed(the pen (/obj/item/pen))
Silent Retard (/mob/living/carbon/human): pointed(the pen (/obj/item/pen))
/datum/callback/verb_callback (/datum/callback/verb_callback): InvokeAsync()
Silent Retard (/mob/living/carbon/human): Point To(the pen (/obj/item/pen))
Silent Retard (/mob/living/carbon/human): Point To(the pen (/obj/item/pen))
Silent Retard (/mob/living/carbon/human): ShiftMiddleClickOn(the pen (/obj/item/pen))
Silent Retard (/mob/living/carbon/human): ClickOn(the pen (/obj/item/pen), "icon-x=14;icon-y=14;middle=1;s...")
the pen (/obj/item/pen): Click(null, "mapwindow.map", "icon-x=14;icon-y=14;middle=1;s...")
the left hand (/atom/movable/screen/inventory/hand): Click(null, "mapwindow.map", "icon-x=14;icon-y=14;middle=1;s...")
Normmacdonaldfan (/client): Click(the left hand (/atom/movable/screen/inventory/hand), null, "mapwindow.map", "icon-x=14;icon-y=14;middle=1;s...")
```

**Comment by LemonInTheDark** [comment]
> Camera consoles (and other embedded views, such as pod launch previews) appear to be extremely broken, uncertain if it's related to this.

I can fix this, but it either means breaking multiz parallax, or forcing a client version update so I can be ensured some byond features work

**Comment by Time-Green** [comment]
> > Camera consoles (and other embedded views, such as pod launch previews) appear to be extremely broken, uncertain if it's related to this.
> 
> I can fix this, but it either means breaking multiz parallax, or forcing a client version update so I can be ensured some byond features work

Force them to update, I wont be able to live without the multiz parallax

**Comment by dragomagol** [comment]
It sounds like this is causing radiation storms/ash storms to be invisible again

**Comment by LemonInTheDark** [comment]
Handled. this still has some bugs tho, with multiz and also camera consoles
workin on em.

**Comment by Baskett55** [comment]
a rod spawned in lower levels of tram and it was invisible, like it was underneath the space tiles, until it reach the station and it looked normal.

**Comment by LemonInTheDark** [comment]
I found another fucking byond bug
this is now number 3 
I hate it here

File: code/game/alternate_appearance.dm [hunk 14]
```diff
@@ -65,14 +65,15 @@ GLOBAL_LIST_EMPTY(active_alternate_appearances)
...
-	var/ghost_appearance
+	var/datum/atom_hud/alternate_appearance/basic/observers/ghost_appearance
 	uses_global_hud_category = FALSE
 
 /datum/atom_hud/alternate_appearance/basic/New(key, image/I, options = AA_TARGET_SEE_APPEARANCE)
 	..()
 	transfer_overlays = options & AA_MATCH_TARGET_OVERLAYS
 	image = I
 	target = I.loc
+	RegisterSignal(target, COMSIG_MOVABLE_Z_CHANGED, .proc/target_z_changed)
```
Comment (ShizCalev) [inline code/game/alternate_appearance.dm:76]: [21:42:15] Runtime in stack_trace.dm, line 3: movable_ztransit overridden. Use override = TRUE to suppress this warning

**Comment by msgerbs** [comment]
I believe #69968 was probably caused by this as well, have you seen it?

**Comment by LemonInTheDark** [comment]
Yeah it was. It's not got a fix PUSHED yet but I've got it

**Comment by LemonInTheDark** [comment]
WAKE ME UP INSIDE https://www.byond.com/forum/post/2821969

**Comment by Time-Green** [comment]
> I'm close to done with this.
> Need to update the pr body and push a few last things, and then we are golden (I hope)
> Oh also test init times so I can make peace with my moth overlord

The future of spaceman is upon us!!!

**Comment by LemonInTheDark** [comment]
There's so much work left to be done brother

**Comment by Time-Green** [comment]
Oh god oh fuck jesus christ it’s happening

**Comment by LemonInTheDark** [comment]
fuck I forgot to update the desc
ah well
plane_cube
